### PR TITLE
[FasterTransformer] Support Unimo-Text

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

New features

### PR changes

Models

### Description

Unimo-Text FasterTransformer supports. 

* Post-normalization supported and verification. 
* `position bias` supported. 
  * During generation, target sequences' position ids start with `len(source sequence)` in dialogue while position ids start with the very beginning 0 in text generation. 
* Make activation alternative.

### CODE REVIEW

**Review by guoshengCS — COMMENTED** [review]
感觉现在最好还是先把加速部分拿出来另外搞个脚本吧，目前千言文本生成的比赛还要依赖这份代码，以防造成困扰

**Comment by FrostML** [comment]
> 感觉现在最好还是先把加速部分拿出来另外搞个脚本吧，目前千言文本生成的比赛还要依赖这份代码，以防造成困扰

Done. Thanks.

File: paddlenlp/ops/faster_transformer/transformer/decoding.py [hunk 1]
```diff
@@ -915,14 +947,6 @@ def __init__(self,
...
                 "linear_weight"] = [self._model.lm_head.decoder_weight.t()]
             self.sub_modules["linear_bias"] = [self._model.lm_head.decoder_bias]
-        self._n_head = self._model.unified_transformer.encoder.layers[
-            0]._config["nhead"]
-        self._hidden_dims = self._model.unified_transformer.encoder.layers[
-            0]._config["d_model"]
-        self._size_per_head = self._hidden_dims // self._n_head
-        self._n_layer = self._model.unified_transformer.encoder.num_layers
-        self._unk_id = self._model.unified_transformer.unk_token_id
-        self._mask_id = self._model.unified_transformer.mask_token_id
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/transformer/decoding.py:925]: 如果是说为了UnifiedTransformer和UnimoText复用InferUnifiedDecoding这里去掉了self._model.unified_transformer并在init另外加入了n_head这些参数的话，可以使用`getattr(self._model, self._model.base_model_prefix)`来获取实际的模型就不用在init另外加参数了，另外这两个模型直接使用`self._model.nhead`这样获取属性应该也是可以的

File: examples/text_generation/unimo-text/faster_unimo/run_gen.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
+import os
+import time
+import argparse
+
```
Comment (smallv0221) [inline examples/text_generation/unimo-text/faster_unimo/run_gen.py]: 这个只有infer的话要不就叫infer.py吧，别叫run_gen了

File: examples/text_generation/unimo-text/faster_unimo/run_gen.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    print('BLEU-4:', bleu4.score())
+
+
+def run(args):
+    paddle.set_device(args.device)
+
+    model = UNIMOLMHeadModel.from_pretrained(args.model_name_or_path)
+    tokenizer = UNIMOTokenizer.from_pretrained(args.model_name_or_path)
+
+    dev_ds = load_dataset(
```
Comment (smallv0221) [inline examples/text_generation/unimo-text/faster_unimo/run_gen.py]: 这里作为一个example，只有infer，但是我们没有直接能用来infer的模型啊

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.h [hunk 4]
```diff
@@ -78,4 +78,7 @@ std::vector<paddle::Tensor> UnifiedDecodingCUDAForward(
     const int& unk_id,
     const int& mask_id,
     const float& temperature,
-    const float& len_penalty);
+    const float& len_penalty,
+    const bool& normalize_before,
+    const bool& pos_bias,
```
Comment (smallv0221) [inline paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.h:83]: 是否可以直接传position_ids呢

File: examples/text_generation/unimo-text/faster_unimo/run_gen.py (hunk 2 above)
Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/run_gen.py]: Done. Thanks.

File: paddlenlp/ops/faster_transformer/transformer/decoding.py (hunk 1 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/transformer/decoding.py:925]: 考虑是当前 API 所涉及的和模型相关的部分仅有参数一项，其他部分最好是传入，以及和之前的 API 调用方式一致，遂如此设置，经讨论，当前 API 可保持原样，修改对应 config 获取方式
Done. Thanks.

File: examples/text_generation/unimo-text/faster_unimo/run_gen.py (hunk 3 above)
Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/run_gen.py]: 加了一个 README 用于说明采用前一路径下的脚本训好的模型。

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.h (hunk 4 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.h:83]: 可以记一个 TODO，目前的情况是，有的模型没有提供 `position id`，而像是生成式 API 又已经提供 ，这里会有两种不同的情况需要分别处理，但是自定义 op 不能支持空的 tensor 或是 vector 的输入，当前的 `pos_bias` 能支持现有的模型，后面看看这里可以怎么完善。

File: examples/text_generation/unimo-text/faster_unimo/README.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+## Python 动态图使用自定义 op
+
+### 编译自定义OP
+
+在 Python 动态图下使用自定义 OP 需要将实现的 C++、CUDA 代码编译成动态库，我们已经提供对应的 CMakeLists.txt ，可以参考使用如下的方式完成编译。同样的自定义 op 编译的说明也可以在自定义 op 对应的路径 `PaddleNLP/paddlenlp/ops/` 下面找到。
+
+#### 克隆 PaddleNLP
+
+首先，因为需要基于当前环境重新编译，当前的 paddlenlp 的 python 包里面并不包含 Faster UNIMO 相关 lib，需要克隆一个 PaddleNLP，并重新编译:
+
```
Comment (guoshengCS) [inline examples/text_generation/unimo-text/faster_unimo/README.md:31]: Python package的paddlenlp路径也可以了

File: examples/text_generation/unimo-text/faster_unimo/README.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+tokenizer = UNIMOTokenizer.from_pretrained(args.model_name_or_path)
+
+model = FasterUnimo(
+    model,
+    decoding_strategy=args.decode_strategy,
+    decoding_lib=args.decoding_lib,
+    use_fp16_decoding=args.use_fp16_decoding)
+```
+
+更详细的例子可以参考 `infer.py`，我们提供了更详细用例。
```
Comment (guoshengCS) [inline examples/text_generation/unimo-text/faster_unimo/README.md]: 说明下当前FT的generate和原生的generate还不是完全一致，只支持基础版本，复杂的策略还不支持吧

Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/README.md]: Done. Thanks.

File: examples/text_generation/unimo-text/faster_unimo/README.md (hunk 5 above)
Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/README.md:31]: Done. Thanks.

**Review by guoshengCS — DISMISSED** [review]

File: examples/text_generation/unimo-text/faster_unimo/README.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
+# Faster UNIMO 预测
+
+在这里我们集成并升级了 Nvidia [Faster Transformer](https://github.com/NVIDIA/FasterTransformer/tree/v3.1) 用于预测加速。以下是使用 Faster UNIMO 的说明。
```
Comment (ZeyuChen) [inline examples/text_generation/unimo-text/faster_unimo/README.md]: NVIDIA，全大写。企业名称要保持准确

File: examples/text_generation/unimo-text/faster_unimo/README.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+其中，
+* `-DSM`: 是指的所用 GPU 的 compute capability。举例来说，可以将之指定为 70(V100) 或是 75(T4)。
+* `-DPY_CMD`: 是指编译所使用的 python，若未指定 `-DPY_CMD` 将会默认使用系统命令 `python` 对应的 Python 版本。
+* `-DWITH_UNIFIED`: 是指是否编译带有 Faster UNIMO 自定义 op 的动态库。
+
+最终，编译会在 `./build/lib/` 路径下，产出 `libdecoding_op.so`，即需要的 Faster UNIMO decoding 执行的库。
+
+## 使用 Faster UNIMO 完成预测
+
+编写 python 脚本的时候，调用 `FasterUnimo` API 并传入 `libdecoding_op.so` 的位置即可实现将 Faster UNIMO 用于当前的预测。
```
Comment (ZeyuChen) [inline examples/text_generation/unimo-text/faster_unimo/README.md]: 这里上下API出现不统一的Unimo大小写拼法，UNIMO/Unimo
是否可以整体统一下。

File: examples/text_generation/unimo-text/faster_unimo/README.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+举例如下：
+
+``` python
+from paddlenlp.ops import FasterUnimo
+
+model = UNIMOLMHeadModel.from_pretrained(args.model_name_or_path)
```
Comment (ZeyuChen) [inline examples/text_generation/unimo-text/faster_unimo/README.md:73]: 上下出现不统一的UNIMO拼写

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py [hunk 10]
```diff
@@ -623,12 +624,187 @@ def __init__(self,
...
+            topp=top_p,
+            max_out_len=max_length,
+            bos_id=self.bos_token_id,
+            eos_id=self.eos_token_id,
+            temperature=temperature,
+            decoding_type_id=decoding_type_id,
+            pos_bias=True)
+
+
+class FasterUnimo(UNIMOPretrainedModel):
```
Comment (ZeyuChen) [inline paddlenlp/ops/faster_transformer/transformer/faster_transformer.py]: 需要快速讨论下你这里的Unimo是保持全大写还是大写开头后小写的形式

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu [hunk 11]
```diff
@@ -226,6 +229,9 @@ std::vector<paddle::Tensor> unified_decoding_kernel(
   decoding_params.type_table =
       reinterpret_cast<const DataType_*>(type_embedding_weight.data<data_t_>());
 
+  ActivationType activate =
+      (hidden_act == "gelu") ? ActivationType::GELU : ActivationType::RELU;
+
   if ("beam_search" == decoding_strategy) {
     TransformerBeamsearch<DecodingTraits_::OpType>*
```
Comment (ZeyuChen) [inline paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu:236]: 从代码整命名上，Beamsearch的形态也出现不一致。理论上作为类名的话，Beam和Search是两个单词，应该独立分开。
TransformerBeamSearch

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu [hunk 12]
```diff
@@ -226,6 +229,9 @@ std::vector<paddle::Tensor> unified_decoding_kernel(
...
         unified_decoding_beamsearch_;
```
Comment (ZeyuChen) [inline paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu]: 然后整套代码中多次出现有的时候_beamsearch, 有的时候beam_search
是原先FT代码遗留的问题，还是我们未统一导致的？

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py [hunk 13]
```diff
@@ -623,12 +624,187 @@ def __init__(self,
...
+                "Topk sampling and topp sampling cannot be both applied. ")
+
+        return self.forward(
+            model_inputs=model_inputs,
+            max_length=max_length,
+            top_k=top_k,
+            top_p=top_p,
+            temperature=temperature)
+
+    def beam_search(self, input_ids, beam_scorer, logits_processors, max_length,
```
Comment (ZeyuChen) [inline paddlenlp/ops/faster_transformer/transformer/faster_transformer.py:788]: 这个beam_search函数是标准的

**Review by ZeyuChen — COMMENTED** [review]
快速对齐下beam_search和UNIMO两个地方的命名风格问题。

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu (hunk 12 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu]: FT 代码里多是使用 `Beamsearch`，`beamsearch` 这种命名方式，所以在 c++ 代码里面是和 FT 统一，而 python API 层面，如下一个 comment 是和 PaddleNLP 内统一

File: examples/text_generation/unimo-text/faster_unimo/README.md (hunk 7 above)
Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/README.md]: Done. Thanks.

File: examples/text_generation/unimo-text/faster_unimo/README.md (hunk 8 above)
Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/README.md]: Done. Thanks. Change to UNIMO

File: examples/text_generation/unimo-text/faster_unimo/README.md (hunk 9 above)
Comment (FrostML) [reply examples/text_generation/unimo-text/faster_unimo/README.md:73]: Done. Thanks.

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py (hunk 10 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/transformer/faster_transformer.py]: Done. Thanks.

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu (hunk 11 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cu:236]: FT 代码里多是使用 Beamsearch，beamsearch 这种命名方式。
原：在 c++ 代码里面是和 FT 统一，而 python API 层面，如下一个 comment 是和 PaddleNLP 内统一。
现：`_beamsearch` 和 `beam_search` 的问题，目前判断是，我们可以先控制我们自己的自定义 op 里面采用一致的写法 `beam_search`，而对于 FT 代码以及 patches 下的内容先保持和 FT 一致。

**Review by smallv0221 — COMMENTED** [review]
LGTM for UNIMO

**Review by ZeyuChen — APPROVED** [review]
LGTM

**Review by guoshengCS — APPROVED** [review]
